dali/pipeline/operators/crop/crop.h

```diff
@@ -75,35 +81,90 @@ class Crop : public SliceBase<Backend>, protected CropAttr {
       case DALI_NFCHW:
         std::tie(F, C, H, W) = std::make_tuple(shape[0], shape[1], shape[2], shape[3]);
         break;
+      case DALI_NDHWC:
+        std::tie(D, H, W, C) = std::make_tuple(shape[0], shape[1], shape[2], shape[3]);
+        break;
+      case DALI_NCDHW:
+        std::tie(C, D, H, W) = std::make_tuple(shape[0], shape[1], shape[2], shape[3]);
+        break;
       default:
-        DALI_FAIL("Not supported layout");
+        DALI_FAIL("Not supported layout[" + std::to_string(layout)
+                  + "] for given number of dimensions" );
     }
 
     auto crop_h = crop_height_[data_idx];
+    if (crop_h <= 0)
+        crop_h = H;
     auto crop_w = crop_width_[data_idx];
-    float anchor_norm[2] = {crop_y_norm_[data_idx], crop_x_norm_[data_idx]};
-    auto crop_anchor = CalculateAnchor(make_span(anchor_norm), {crop_h, crop_w}, {H, W});
-    int64_t crop_y = crop_anchor[0], crop_x = crop_anchor[1];
+    if (crop_w <= 0)
+        crop_w = W;
 
-    switch (layout) {
-      case DALI_NHWC:
-        slice_anchors_[data_idx] = {crop_y, crop_x, 0};
-        slice_shapes_[data_idx] = {crop_h, crop_w, C};
-        break;
-      case DALI_NCHW:
-        slice_anchors_[data_idx] = {0, crop_y, crop_x};
-        slice_shapes_[data_idx] = {C, crop_h, crop_w};
-        break;
-      case DALI_NFHWC:
-        slice_anchors_[data_idx] = {0, crop_y, crop_x, 0};
-        slice_shapes_[data_idx] = {F, crop_h, crop_w, C};
-        break;
-      case DALI_NFCHW:
-        slice_anchors_[data_idx] = {0, 0, crop_y, crop_x};
-        slice_shapes_[data_idx] = {F, C, crop_h, crop_w};
-        break;
-      default:
-        DALI_FAIL("Not supported layout");
+    if (is_volumetric_layout) {
```

Comment:
It's get really complicated below, would be nice if in the layout switch above you defined `anchor_norm`, and `crop_window` in layout agnostic way
So for example:
```
case DALI_NFCHW:
        std::tie(F, C, H, W) = std::make_tuple(shape[0], shape[1], shape[2], shape[3]);
		float anchor_norm[4] =
        {0, 0, crop_y_norm_[data_idx], crop_x_norm_[data_idx]};
      float crop_window[4] = {F, C, crop_h, crop_w};
        break;
```

Than you would just
```
 slice_anchors_[data_idx] = CalculateAnchor(make_span(anchor_norm), crop_window, input_shape);
slice_shapes_[data_idx] = crop_window;
```

Btw, why don't we use CropWindowGenerator here?

Reply:
good catch. Will refactor this to use CropWindowGenerator